Add mCaptcha support to Mastodon

lenikadali opened the pull request:
Fixes #32569

lenikadali committed:
Initial commit for adding mCaptcha support

Initial commit for adding mCaptcha support

The implementation will need cleaning up
as verification of the behaviour being what we want
requires testing in a staging environment as
the development environment doesn't provide a user
sign up flow where a user would solve a CAPTCHA
(if it is enabled on the server) like it would
on production

lenikadali commented:
The implementation will need cleaning up as verification of the behaviour being what we want most likely requires testing in a staging environment. 

The development environment doesn't provide a user sign up flow where a user would be prompted to solve a CAPTCHA
(if it is enabled on the server) like it would on production.

lenikadali reviewed (commented):
The files in `mcaptcha/` are intended to mirror the implementation of the [hcaptcha](https://github.com/Nexus-Mods/hcaptcha/) gem without the gem specific parts (e.g. the `gemspec` file). I opted for this instead of making a new gem and needing to get it registered on RubyGems.

I noticed that despite having been funded by NLnet, the [mCaptcha project](https://github.com/mCaptcha/mCaptcha) hasn't received updates in a while. Another captcha alternative I came across recently is [ALTCHA](https://github.com/altcha-org/altcha) which may be a better fit. If we opt to switch to ALTCHA, it may be easier to use [their Ruby library](https://github.com/altcha-org/altcha-lib-rb)

lenikadali committed:
Update confirm_captcha check to use either method

Updated confirm_captcha check to use either method
available depending on whether it is mCaptcha or hCaptcha
enabled.

Currently doesn't work since we need to change it
to be compatible with an if/elsif statement. Adding it now
so that we don't forget update it during cleanup/changes
related to feedback from testing.

lenikadali commented on app/controllers/auth/confirmations_controller.rb:
```diff
@@ -29,10 +29,16 @@ def new
-      flash.now[:alert] = message
-      render :captcha
-      return
+    if Rails.configuration.x.captcha.mcaptcha_secret_key.present?
+      check_mcaptcha! do |message|
+        flash.now[:alert] = message
+        render :captcha
+        return
+    else
+      check_mcaptcha! do |message|
+        flash.now[:alert] = message
+        render :captcha
+        return
```
This currently fails RuboCop with the following trace:

```
F: Lint/Syntax: else without rescue is useless
(Using Ruby 3.2 parser; configure using TargetRubyVersion parameter, under AllCops)
    else
    ^^^^
app/controllers/auth/confirmations_controller.rb:103:1: F: Lint/Syntax: unexpected token $end
(Using Ruby 3.2 parser; configure using TargetRubyVersion parameter, under AllCops)
```

Tried to move things around and it didn't look to work. Probably relies on re-working the config check as mentioned [here](https://github.com/mastodon/mastodon/pull/33524#discussion_r1908832995)

ClearlyClaire replied:
There is a syntax error here, you're missing an `end` to the `do` block.

lenikadali replied:
Thanks for catching this :raised_hands:

vmstan commented:
Thank you for taking this on.

lenikadali committed:
Merge branch 'main' into feature/add-mcaptcha-support

lenikadali committed:
Add missing end keyword

lenikadali committed:
Remove unneeded true/false, add missing module

Added the missing ViewMethods module and removed
the true/false statements from the Auth::CaptchaConcern
which weren't needed

lenikadali commented on app/controllers/concerns/auth/captcha_concern.rb:
```diff
@@ -4,13 +4,22 @@ module Auth::CaptchaConcern
+      true
+    else
+      false
+    end
```
Have removed the unneeded `true` and `false` statements but will otherwise leave as is since the way we check for the secret key is derived from how the configuration is setup and referenced.

lenikadali commented on config/captcha.yml:
```diff
@@ -1,3 +1,5 @@
 shared:
-  secret_key: <%= ENV.fetch('HCAPTCHA_SECRET_KEY', nil) %>
-  site_key: <%= ENV.fetch('HCAPTCHA_SITE_KEY', nil) %>
+  hcaptcha_secret_key: <%= ENV.fetch('HCAPTCHA_SECRET_KEY', nil) %>
+  hcaptcha_site_key: <%= ENV.fetch('HCAPTCHA_SITE_KEY', nil) %>
+  mcaptcha_secret_key: <%= ENV.fetch('MCAPTCHA_SECRET_KEY', nil) %>
+  mcaptcha_site_key: <%= ENV.fetch('MCAPTCHA_SITE_KEY', nil) %>
```
[This comment](https://github.com/mastodon/mastodon/pull/33524#discussion_r1924149229) applies here as well

ClearlyClaire commented:
Hi! Thank you for your contribution!

While I am not entirely convinced with mCaptcha, we are not opposed to offer it as an alternative to hCaptcha, which does have its drawbacks.

I have not reviewed the PR yet, as we usually expect the automated checks to pass before doing so, and a bunch have failed here. They *should* be providing enough information for you to address the issues, but feel free to ask if you need help understanding or addressing them.

Also, it may make sense for you to run those checks locally for quicker feedback. The “Ruby linting” one is basically just running `bundle exec rubocop`, though the “Ruby testing” ones are more involved. To run the latter, set `RAILS_ENV` to `tests`, build assets (`bundle exec rails assets:precompile`), reset the database (`bundle exec rails db:reset`) and then run the tests themselves `bundle exec rspec`.

lenikadali commented:
> Hi! Thank you for your contribution!
> 
> While I am not entirely convinced with mCaptcha, we are not opposed to offer it as an alternative to hCaptcha, which does have its drawbacks.
> 
> I have not reviewed the PR yet, as we usually expect the automated checks to pass before doing so, and a bunch have failed here. They _should_ be providing enough information for you to address the issues, but feel free to ask if you need help understanding or addressing them.
> 
> Also, it may make sense for you to run those checks locally for quicker feedback. The “Ruby linting” one is basically just running `bundle exec rubocop`, though the “Ruby testing” ones are more involved. To run the latter, set `RAILS_ENV` to `tests`, build assets (`bundle exec rails assets:precompile`), reset the database (`bundle exec rails db:reset`) and then run the tests themselves `bundle exec rspec`.

Thanks so much for the feedback 🙌
This PR is still a draft because I have to fix the linting and (any) failing tests so I wasn't expecting a formal review until I remove the draft status. I hope that's okay 🙏

Also I'll need to do end-to-end tests for this so I anticipate this taking some time.

lenikadali committed:
Merge branch 'main' into feature/add-mcaptcha-support

renchap commented:
> Also I'll need to do end-to-end tests for this so I anticipate this taking some time.

I suspect that end-to-end tests will be hard to implement with an external service like this, and we do not have a lot of them in place right now, so do not spend too much time on this, you can submit it without them if that is too hard to implement.

lenikadali committed:
Merge branch 'main' into feature/add-mcaptcha-support

lenikadali committed:
Partial fix of .html_safe usage

lenikadali committed:
RuboCop fixes, revert tag.div change

Added RuboCop fixes and reverted the tag.div change
since there was no (apparent) way to make it a 1:1 with
the previous implementation that uses string interpolation

lenikadali committed:
Re-order files to fit app convention

Re-ordered the files to fit app convention.
Hopefully should fix the pipeline failure
regarding the inability to find the ViewMethods
for MCaptcha.

lenikadali committed:
Merge branch 'main' into feature/add-mcaptcha-support

lenikadali committed:
Rename error class, add ::Mcaptcha

Renamed the error class and added a new
include line ending with ::Mcaptcha to see
if that solves the unitialized constant error
Auth::CaptchaConcern::Mcaptcha since we've *definitely*
included the Mcaptcha::Adapters::ViewMethods

lenikadali committed:
Add remaining files from hcaptcha gem implementation

Added the remaining files from the hcaptcha gem implementation
in the hope that it will make it easier to solve the inability
to find Mcaptcha::Adapters::ViewMethods

lenikadali committed:
Merge branch 'main' into feature/add-mcaptcha-support

lenikadali committed:
Move mcaptcha files to app/lib

Moved mcaptcha files to app/lib to solve the inability
to detect ViewMethods

lenikadali committed:
Merge branch 'main' into feature/add-mcaptcha-support

lenikadali commented:
Going to close this because the debug process is a bit too brittle (have been bitten by webpack interactions and I think it is too noisy to rely on the tests being run on GitHub before seeing what the new problem should be)
